crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -639,3 +655,301 @@ fn normalize_string(
+fn count_indentation_like_black(line: &str) -> TextSize {
+    let tab_width: u32 = 8;
+    let mut indentation = TextSize::default();
+    for char in line.chars() {
+        if char == '\t' {
+            // Pad to the next multiple of tab_width
+            indentation += TextSize::from(tab_width - (indentation.to_u32().rem_euclid(tab_width)));
+        } else if char.is_whitespace() {
+            indentation += char.text_len();
+        } else {
+            return indentation;
+        }
+    }
+    indentation
+}
+
+/// Format a docstring by trimming whitespace and adjusting the indentation.
+///
+/// Summary of changes we make:
+/// * Normalize the string like all other strings
+/// * Ignore docstring that have an escaped newline
+/// * Trim all trailing whitespace, except for a chaperone space to avoid quotes or backslashes
+///   in the last line.
+/// * Trim leading whitespace on the first line, again except for a chaperone space
+/// * If there is only content in the first line and after that only whitespace, collapse the
+///   docstring into one line
+/// * Adjust the indentation (see below)
+///
+/// # Docstring indentation
+///
+/// Unlike any other string, like black we change the indentation of docstring lines.
+///
+/// We want to preserve the indentation inside the docstring relative to the suite statement/block
+/// indent that the docstring statement is in, but also want to apply the change of the outer
+/// indentation in the docstring, e.g.
+/// ```python
+/// def sparkle_sky():
+///   """Make a pretty sparkly sky.
+///   *       * ✨        *.    .
+///      *       *      ✨      .
+///      .  *      . ✨    * .  .
+///   """
+/// ```
+/// should become
+/// ```python
+/// def sparkle_sky():
+///     """Make a pretty sparkly sky.
+///     *       * ✨        *.    .
+///        *       *      ✨      .
+///        .  *      . ✨    * .  .
+///     """
+/// ```
+/// We can't compute the full indentation here since we don't know what the block indent of
+/// the doc comment will be yet and which we can only have added by formatting each line
+/// separately with a hard line break. This means we need to strip shared indentation from
+/// docstring while preserving the in-docstring bigger-than-suite-statement indentation. Example:
+/// ```python
+/// def f():
+///  """first line
+///  line a
+///     line b
+///  """
+/// ```
+/// The docstring indentation is 2, the block indents will change this to 4 (but we can't
+/// determine this at this point). The indentation of line a is 2, so we trim `  line a`
+/// to `line a`. For line b it's 5, so we trim it to `line b` and pad with 5-2=3 spaces to
+/// `   line b`. The closing quotes, being on their own line, are stripped get only the
+/// default indentation. Fully formatted:
+/// ```python
+/// def f():
+///    """first line
+///    line a
+///       line b
+///    """
+/// ```
+///
+/// Tabs are counted by padding them to the next multiple of 8 according to `str.expandtabs`. When
```

Comment:
```suggestion
/// Tabs are counted by padding them to the next multiple of 8 according to `str.expandtabs`. When
```

Add a link to `str_expandtabs`. It's otherwise unclear what it refers to